prost-types/src/lib.rs

```diff
@@ -165,6 +169,39 @@ impl Timestamp {
         // debug_assert!(self.seconds >= -62_135_596_800 && self.seconds <= 253_402_300_799,
         //               "invalid timestamp: {:?}", self);
     }
+
+    /// Creates a new `Timestamp` at the start of the provided UTC date.
+    pub fn date(year: i64, month: u8, day: u8) -> Timestamp {
+        Timestamp::date_time_nanos(year, month, day, 0, 0, 0, 0)
+    }
+
+    /// Creates a new `Timestamp` instance with the provided UTC date and time.
+    pub fn date_time(year: i64, month: u8, day: u8, hour: u8, minute: u8, second: u8) -> Timestamp {
+        Timestamp::date_time_nanos(year, month, day, hour, minute, second, 0)
+    }
+
+    /// Creates a new `Timestamp` instance with the provided UTC date and time.
+    pub fn date_time_nanos(
+        year: i64,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+        nanos: u32,
+    ) -> Timestamp {
+        let date_time = datetime::DateTime {
+            year,
+            month,
+            day,
+            hour,
+            minute,
+            second,
+            nanos,
+        };
+        assert!(date_time.is_valid(), "invalid date time: {}", date_time);
```

Comment:
I not a huge fan of these potential panics, should we make them an error?

Reply:
Yeah I'm in favor of making this an error. This PR already introduces one new error type, `TimestampParseError`, and there is already an existing `TimestampOutOfSystemRangeError`, and this would be an additional.  WDYT about folding all of these cases into an enum of 3 variants, instead of separate types?  That would be a breaking change, but I imagine there aren't many users of `TimestampOutOfSystemRangeError` so it'd hopefully not be too painful of an upgrade for folks.

Reply:
(for a bit more context, subsequent PRs are likely going to require a new version bump anyway because I'm planning on adding default trait impls for generated types, which is technically a breaking change as well)

Reply:
I've gone ahead and pushed this as the latest commit

Reply:
Yeah I am happy to make that breaking change.